# Step 1 of the task

## Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Importing Data

In [2]:
path = r'C:\Users\ramis\OneDrive\JC\Data Analytics Course\Data Analytics Immersion\A4\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\ramis\\OneDrive\\JC\\Data Analytics Course\\Data Analytics Immersion\\A4\\Instacart Basket Analysis'

In [4]:
# import the orders_products_merged_2.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_2.pkl'))

In [5]:
# check output
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days         Fewest orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days         Fewest orders  
4      Least busy    Slowest days           Most orders

In [6]:
# create a subset of the firt million rows of the dataframe
df = ords_prods_merge[:1000000]

In [7]:
df.shape

(1000000, 18)

In [8]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days         Fewest orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days         Fewest orders  
4      Least busy    Slowest days           Most orders

In [9]:
df.groupby('product_name')

In [10]:
# group by department_id and agg on order_number
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [11]:
# Performing Multiple Aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Step 4 of the task: creating a loyalty flag for existing customers using the transform() and loc() functions.

In [12]:
# Aggregating Data with transform(): Split the data into groups, apply the transform() function, and create a new column, "max_order"
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [13]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range produ

In [14]:
pd.options.display.max_rows = None

In [15]:
# Deriving Columns with loc() to create the loyalty_flag
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
# checking frequency of the new "loyalty_flag" column
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [18]:
# check the output on specific columns
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

### As discussed with my tutor, there are rows where the prices = 99999. They are row with product_id= 33664. I will change the values of the respective cells to 9.9 before I proceed further with the task of this exercise.

In [25]:
# Count the number of rows where product_id is equal to 33664
product_count = len(ords_prods_merge[ords_prods_merge['product_id'] == 33664])

# Count the total number of rows in the DataFrame
total_rows = len(ords_prods_merge)

# Calculate the percentage and round it to four decimal places
percentage = (product_count / total_rows) * 100
percentage_rounded = round(percentage, 4)

print("The percentage of rows where product_id = 33664 is", percentage_rounded, "%")

The percentage of rows where product_id = 33664 is 0.0022 %


In [31]:
# Update 'prices' where 'product_id' is 33664 to 9.9
ords_prods_merge.loc[ords_prods_merge['product_id'] == 33664, 'prices'] = 9.9

In [33]:
# verify the change in the max of prices as it was 99999
ords_prods_merge.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.240486e+07  3.240486e+07  3.240486e+07        3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01        2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01        2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00        1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01        3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01        6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110408e+01  2.559866e+04   
std         4.246380e+00            8.779064e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices     max_order  
count  3.240486e+07  3.240486e+07  
mean   9.826462e+00  3.305217e+01  
std    1.741430e+02  2.515525e+01  
min    1.000000e+00  1.000000e+00  
25%    4.200000e+00  1.300000e+01  
50%    7.400000e+00  2.600000e+01  
75%    1.130000e+01  4.700000e+01  
max    1.490000e+04  9.900000e+01

In [36]:
# checking rows where the product_id = 33664 by using the head() function
filtered_rows = ords_prods_merge[ords_prods_merge['product_id'] == 33664]
print(filtered_rows.head())

          order_id  user_id  order_number  orders_day_of_week  \
29165516    183964      873             3                   0   
29165517   1851256      873             4                   6   
29165518   1915696     1893             1                   5   
29165519   2763293     1893             2                   4   
29165520   2564805     1893             4                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
29165516                 10                     7.0       33664   
29165517                 12                    13.0       33664   
29165518                 17                     NaN       33664   
29165519                 16                    13.0       33664   
29165520                 17                    30.0       33664   

          add_to_cart_order  reordered           product_name  aisle_id  \
29165516                 11          0  2 % Reduced Fat  Milk        84   
29165517                  8          1  2 % Reduced Fat 

# Step 2 of the task

In [37]:
# create a panda object with the groupby() function
ords_prods_merge.groupby('product_name')

In [38]:
# group by department_id and agg order_number
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3 of the task: to start with, the output of the entire dataframe contains the entire range of department_id's 1-21. In addition, the means, when computed for the entire dataset, exhibit lower variation compared to when calculated for a smaller subset.

# Step 5 of the task

In [40]:
# Group the data by 'loyalty_flag' and calculate basic statistics for 'prices'
loyalty_stats = ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

# Print the statistics for each loyalty category
print(loyalty_stats)

                       count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0   9.579352  164.031161  1.0  4.2  7.4  11.2   
New customer       6243990.0  10.059912  183.437157  1.0  4.2  7.4  11.3   
Regular customer  15876776.0   9.894716  176.730636  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    14900.0  
New customer      14900.0  
Regular customer  14900.0  


## the output above shows the the mean of prices for the three loyalty groups is almost the same. they are 9.57, 10.05 and 9.89 for Loyal customer, New customer, and Regular customer respectively. This suggests that the prices of products purchased by loyal customers do not  differ from those purchased by regular or new customers.

# Step 6 of the task

In [41]:
# Calculate the mean of prices for each user and create a new column 'avg_price'
ords_prods_merge['avg_price'] = ords_prods_merge.groupby('user_id')['prices'].transform('mean')

In [42]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  department_id  prices  _merge    price_range_loc  \
0          Soda  ...              7     9.0    both  Mid-range product   
1          Soda  ...              7     9.0    both  Mid-range product   
2          Soda  ...              7     9.0    both  Mid-range product   
3          Soda  ...              7     9.0    both  Mid-range product   
4          Soda  ...              7     9.0    both  Mid-range product   
5          Soda  ...              7     9.0    both  Mid-range product   
6          Soda  ...              7     9.0    both  Mid-range product   
7          Soda  ...              7     9.0    both  Mid-range product   
8          Soda  ...              7     9.0    both  Mid-range product   
9          Soda  ...              7     9.0    both  Mid-range product   
10         Soda  ...              7   

In [43]:
# Create a spending flag based on the average price
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [47]:
ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0         Soda  ...     9.0    both  Mid-range product  Regularly busy   
1         Soda  ...     9.0    both  Mid-range product  Regularly busy   
2         Soda  ...     9.0    both  Mid-range product  Regularly busy   
3         Soda  ...     9.0    both  Mid-range product      Least busy   
4         Soda  ...     9.0    both  Mid-range product      Least busy   
5         Soda  ...     9.0    both  Mid-range product  Regularly busy   
6         Soda  ...     9.0    both  Mid-range product  Regularly busy   
7         Soda  ...     9.0    both  Mid-range product  Regularly busy   
8         Soda  ...     9.0    both  Mid-range product  Regularly busy   
9         Soda  ...     9.0    both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag  avg_price  \
0  Regularly busy        Average orders        10  New customer   6.367797   
1    Slowest days         Fewest orders        10  New customer   6.367797   
2    Slowest days           Most orders        10  New customer   6.367797   
3    Slowest days         Fewest orders        10  New customer   6.367797   
4    Slowest days           Most orders        10  New customer   6.367797   
5  Regularly busy         Fewest orders        10  New customer   6.367797   
6    Busiest days        Average orders        10  New customer   6.367797   
7    Busiest days           Most orders        10  New customer   6.367797   
8    Busiest days           Most orders        10  New customer   6.367797   
9    Slowest days        Average orders        10  New customer   6.367797   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  
5   Low spender  
6   Low spender  
7   Low spender  
8   Low spender  
9   Low spender  

[10 rows x 22 columns]

In [48]:
# checking frequency of the new "spending_flag" column
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31828060
High spender      576799
Name: count, dtype: int64

# Step 7 of the task

In [49]:
# Calculate the median of 'days_since_prior_order' for each user and create a new column 'median_days_since_prior_order'
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform('median')

In [50]:
# Create an order frequency flag based on the median
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] > 10) & (ords_prods_merge['median_days_since_prior_order'] <= 20), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [51]:
ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day    busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy    Slowest days   
2         Soda  ...  Mid-range product  Regularly busy    Slowest days   
3         Soda  ...  Mid-range product      Least busy    Slowest days   
4         Soda  ...  Mid-range product      Least busy    Slowest days   
5         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
6         Soda  ...  Mid-range product  Regularly busy    Busiest days   
7         Soda  ...  Mid-range product  Regularly busy    Busiest days   
8         Soda  ...  Mid-range product  Regularly busy    Busiest days   
9         Soda  ...  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag avg_price spending_flag  \
0        Average orders        10  New customer  6.367797   Low spender   
1         Fewest orders        10  New customer  6.367797   Low spender   
2           Most orders        10  New customer  6.367797   Low spender   
3         Fewest orders        10  New customer  6.367797   Low spender   
4           Most orders        10  New customer  6.367797   Low spender   
5         Fewest orders        10  New customer  6.367797   Low spender   
6        Average orders        10  New customer  6.367797   Low spender   
7           Most orders        10  New customer  6.367797   Low spender   
8           Most orders        10  New customer  6.367797   Low spender   
9        Average orders        10  New customer  6.367797   Low spender   

   median_days_since_prior_order   order_frequency_flag  
0                           20.5  Non-frequent customer  
1                           20.5  Non-frequent customer  
2                           20.5  Non-frequent customer  
3                           20.5  Non-frequent customer  
4                           20.5  Non-frequent customer  
5                           20.5  Non-frequent customer  
6                           20.5  Non-frequent customer  
7                           20.5  Non-frequent customer  
8                           20.5  Non-frequent customer  
9                           20.5  Non-frequent customer  

[10 rows x 24 columns]

In [52]:
# checking frequency of the new "order_frequency_flag" column
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [53]:
# investigating order_frequency_flag with NaN values:
# Filter rows where 'order_frequency_flag' is NaN
rows_with_nan = ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()]

# Print the filtered rows
print(rows_with_nan)

          order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0

### as per the output above, it seems that this user has only one order. Thus, we cannot calculate median_days_since_prior_order as there is only one single order. Therefore, we could consider updating the 5 respective NaN order_frequency_flag's to Non-frequent customer.


# Step 8 of the task: check done.

# Step 9 of the task

In [54]:
# Export data to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_3.pkl'))